### PR TITLE
Cache results of sub-repo perms enabled checks

### PR DESCRIPTION
**Pull request by pjlast**
Moves the SubRepoPermsEnabledForRepo checks back to the top level function so that we can cache the results on a repo level.

## Test plan

Tests should still pass.

### CODE REVIEW

File: internal/search/job/jobutil/sub_repo_perms_job.go [hunk 1]
```diff
@@ -84,7 +85,28 @@ func applySubRepoFiltering(ctx context.Context, checker authz.SubRepoPermissionC
...
+		// Skip check if sub-repo perms are disabled for the repository
+		enabled, ok := subRepoPermsCache[m.RepoName().Name]
+		if ok && !enabled {
+			filtered = append(filtered, m)
+			continue
+		}
+		if !ok {
+			enabled, err := authz.SubRepoEnabledForRepo(ctx, checker, m.RepoName().Name)
+			if err != nil {
+				// If an error occurs while checking sub-repo perms, we ommit it from the results
```
Comment (sashaostrikov) [inline internal/search/job/jobutil/sub_repo_perms_job.go:100]: typo fix:
```suggestion
				// If an error occurs while checking sub-repo perms, we omit it from the results
```

File: internal/search/job/jobutil/sub_repo_perms_job.go [hunk 2]
```diff
@@ -84,7 +85,28 @@ func applySubRepoFiltering(ctx context.Context, checker authz.SubRepoPermissionC
...
+
 	for _, m := range matches {
+		// Skip check if sub-repo perms are disabled for the repository
+		enabled, ok := subRepoPermsCache[m.RepoName().Name]
+		if ok && !enabled {
+			filtered = append(filtered, m)
+			continue
+		}
+		if !ok {
+			enabled, err := authz.SubRepoEnabledForRepo(ctx, checker, m.RepoName().Name)
```
Comment (sashaostrikov) [inline internal/search/job/jobutil/sub_repo_perms_job.go:98]: do we want to cache errors as well? 🤔

**Review by sashaostrikov — APPROVED** [review]
LGTM!

File: internal/search/job/jobutil/sub_repo_perms_job.go (hunk 2 above)
Comment (pjlast) [reply internal/search/job/jobutil/sub_repo_perms_job.go:98]: Good question 🤔

**Review by keegancsmith — APPROVED** [review]
LGTM. This pattern seems useful in other places, could it be applied elsewhere? If so there is likely a nice little struct hiding most of these implementation details making the important bits of business logic much clearer. Not for this PR though.

**Pull request merged**